lang/java/avro/src/main/java/org/apache/avro/generic/GenericData.java

```diff
@@ -94,24 +95,38 @@ public GenericData(ClassLoader classLoader) {
   /** Return the class loader that's used (by subclasses). */
   public ClassLoader getClassLoader() { return classLoader; }
 
-  public Map<String, Conversion<?>> conversions =
+  private Map<String, Conversion<?>> conversions =
       new HashMap<String, Conversion<?>>();
 
-  public Map<Class<?>, Conversion<?>> conversionsByClass =
-      new IdentityHashMap<Class<?>, Conversion<?>>();
+  private Map<Class<?>, List<Conversion<?>>> conversionsByClass =
+      new IdentityHashMap<Class<?>, List<Conversion<?>>>();
+
+  public Iterable<Conversion<?>> getAllConversions() {
+    return conversions.values();
+  }
 
   public void addLogicalTypeConversion(Conversion<?> conversion) {
     conversions.put(conversion.getLogicalTypeName(), conversion);
-    conversionsByClass.put(conversion.getConvertedType(), conversion);
+    Class<?> type = conversion.getConvertedType();
+    if (conversionsByClass.containsKey(type)) {
```

Comment:
Since GenericData is usually a singleton, shouldn't it be threadsafe?

These changes aren't, but AFAICT theogical type stuff prior to this patch isn't either.